builtin/providers/librato/resource_librato_metric_test.go

```diff
@@ -0,0 +1,230 @@
+					resource.TestCheckResourceAttr(
+						"librato_metric.foobar", "name", name),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckLibratoMetricDestroy(s *terraform.State) error {
+	client := testAccProvider.Meta().(*librato.Client)
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "librato_metric" {
+			continue
+		}
+
+		fmt.Printf("*** testAccCheckLibratoMetricDestroy ID: %s\n", rs.Primary.ID)
+		_, _, err := client.Metrics.Get(rs.Primary.ID)
+		if err == nil {
+			return fmt.Errorf("Metric still exists")
+		}
+	}
+
+	return nil
+}
+
+func testAccCheckLibratoMetricName(metric *librato.Metric, name string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		if metric.Name == nil || *metric.Name != name {
+			return fmt.Errorf("Bad name: %s", *metric.Name)
+		}
+
+		return nil
+	}
+}
+
+func testAccCheckLibratoMetricType(metric *librato.Metric, validTypes []string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		m := make(map[string]bool)
+		for _, v := range validTypes {
+			m[v] = true
+		}
+
+		if !m[*metric.Type] {
+			return fmt.Errorf("Bad metric type: %s", *metric.Type)
+		}
+
+		return nil
+	}
+}
+
+func testAccCheckLibratoMetricExists(n string, metric *librato.Metric) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[n]
+
+		if !ok {
+			return fmt.Errorf("Not found: %s", n)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("No Metric ID is set")
+		}
+
+		client := testAccProvider.Meta().(*librato.Client)
+
+		foundMetric, _, err := client.Metrics.Get(rs.Primary.ID)
+
+		if err != nil {
+			return err
+		}
+
+		if foundMetric.Name == nil || *foundMetric.Name != rs.Primary.ID {
+			return fmt.Errorf("Metric not found")
+		}
+
+		*metric = *foundMetric
+
+		return nil
+	}
+}
+
+func counterMetricConfig(name, typ, desc string) string {
+	return strings.TrimSpace(fmt.Sprintf(`
+    resource "librato_metric" "foobar" {
+        name = "%s"
+        type = "%s"
+        description = "%s"
+        attributes {
+          display_stacked = true,
+          created_by_ua = "go-librato/0.1"
+        }
+    }`, name, typ, desc))
+}
+
+func gaugeMetricConfig(name, typ, desc string) string {
+	return strings.TrimSpace(fmt.Sprintf(`
+    resource "librato_metric" "foobar" {
+        name = "%s"
+        type = "%s"
+        description = "%s"
+        attributes {
+          display_stacked = true,
+          created_by_ua = "go-librato/0.1"
+        }
+    }`, name, typ, desc))
+}
+
+func compositeMetricConfig(name, typ, desc string) string {
+	return strings.TrimSpace(fmt.Sprintf(`
+    resource "librato_metric" "foobar" {
+        name = "%s"
+        type = "%s"
+        description = "%s"
+        composite = "s(\"librato.cpu.percent.user\", {\"environment\" : \"prod\", \"service\": \"api\"})"
+        attributes {
+          display_stacked = true,
+          created_by_ua = "go-librato/0.1"
```

Comment:
The tests fail w/o this in the config. Perhaps I'm taking the wrong approach in fixing the issue:

```
hashicorp/terraform - [cj-add-librato-metrics●] » TF_LOG=info make testacc TEST=./builtin/providers/librato TESTARGS="-run=TestAccLibratoCompositeMetric"
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/05/18 14:45:56 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/librato -v -run=TestAccLibratoCompositeMetric -timeout 120m
=== RUN   TestAccLibratoCompositeMetric
2017/05/18 14:45:59 [WARN] Test: Executing step 0
2017/05/18 14:45:59 [INFO] terraform: building graph: GraphTypeValidate
2017/05/18 14:45:59 [WARN] terraform: shadow graph disabled
2017/05/18 14:45:59 [INFO] terraform: building graph: GraphTypeRefresh
2017/05/18 14:45:59 [WARN] terraform: shadow graph disabled
2017/05/18 14:45:59 [INFO] terraform: building graph: GraphTypePlan
2017/05/18 14:45:59 [WARN] terraform: shadow graph disabled
2017/05/18 14:45:59 [WARN] Test: Step plan: DIFF:

CREATE: librato_metric.foobar
  attributes.#:                 "0" => "1"
  attributes.0.display_stacked: "" => "true"
  composite:                    "" => "s(\"librato.cpu.percent.user\", {\"environment\" : \"prod\", \"service\": \"api\"})"
  description:                  "" => "A test composite metric"
  name:                         "" => "tftest-metric-nqvxkppkjs"
  type:                         "" => "composite"

STATE:

<no state>
2017/05/18 14:45:59 [INFO] terraform: building graph: GraphTypeApply
2017/05/18 14:45:59 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypePlan
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypeRefresh
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [INFO] Reading Librato Metric: tftest-metric-nqvxkppkjs
2017/05/18 14:46:00 [INFO] Read Librato Metric: {"name":"tftest-metric-nqvxkppkjs","description":"A test composite metric","type":"composite","composite":"s(\"librato.cpu.percent.user\", {\"environment\" : \"prod\", \"service\": \"api\"})","attributes":{"color":null,"display_max":null,"display_min":null,"display_units_long":"","display_units_short":"","display_stacked":true,"display_transform":"","created_by_ua":"go-librato/0.1"}}
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypePlan
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [WARN] Test: Executing destroy step
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypeValidate
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypeRefresh
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [INFO] Reading Librato Metric: tftest-metric-nqvxkppkjs
2017/05/18 14:46:00 [INFO] Read Librato Metric: {"name":"tftest-metric-nqvxkppkjs","description":"A test composite metric","type":"composite","composite":"s(\"librato.cpu.percent.user\", {\"environment\" : \"prod\", \"service\": \"api\"})","attributes":{"color":null,"display_max":null,"display_min":null,"display_units_long":"","display_units_short":"","display_stacked":true,"display_transform":"","created_by_ua":"go-librato/0.1"}}
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypePlanDestroy
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [WARN] Test: Step plan: DIFF:

DESTROY: librato_metric.foobar

STATE:

librato_metric.foobar:
  ID = tftest-metric-nqvxkppkjs
  attributes.# = 1
  attributes.0.aggregate = false
  attributes.0.color = 
  attributes.0.created_by_ua = go-librato/0.1
  attributes.0.display_max = 
  attributes.0.display_min = 
  attributes.0.display_stacked = true
  attributes.0.display_units_long = 
  attributes.0.display_units_short = 
  attributes.0.gap_detection = false
  composite = s("librato.cpu.percent.user", {"environment" : "prod", "service": "api"})
  description = A test composite metric
  name = tftest-metric-nqvxkppkjs
  type = composite
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypeApply
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [INFO] Deleting Metric: tftest-metric-nqvxkppkjs
2017/05/18 14:46:00 [INFO] Verifying Metric tftest-metric-nqvxkppkjs deleted
2017/05/18 14:46:00 [INFO] GETing Metric tftest-metric-nqvxkppkjs
2017/05/18 14:46:00 [INFO] GET returned a 404 for Metric tftest-metric-nqvxkppkjs
2017/05/18 14:46:00 [INFO] I think Metric is deleted
*** testAccCheckLibratoMetricDestroy ID: tftest-metric-nqvxkppkjs
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypePlanDestroy
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypeRefresh
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypePlanDestroy
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
--- FAIL: TestAccLibratoCompositeMetric (0.92s)
	testing.go:280: Step 0 error: After applying this step and refreshing, the plan was not empty:
		
		DIFF:
		
		UPDATE: librato_metric.foobar
		  attributes.0.created_by_ua: "go-librato/0.1" => ""
hashicorp/terraform - [cj-add-librato-metrics●] » TF_LOG=info make testacc TEST=./builtin/providers/librato TESTARGS="-run=TestAccLibratoCompositeMetric"
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/05/18 14:45:56 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/librato -v -run=TestAccLibratoCompositeMetric -timeout 120m
=== RUN   TestAccLibratoCompositeMetric
2017/05/18 14:45:59 [WARN] Test: Executing step 0
2017/05/18 14:45:59 [INFO] terraform: building graph: GraphTypeValidate
2017/05/18 14:45:59 [WARN] terraform: shadow graph disabled
2017/05/18 14:45:59 [INFO] terraform: building graph: GraphTypeRefresh
2017/05/18 14:45:59 [WARN] terraform: shadow graph disabled
2017/05/18 14:45:59 [INFO] terraform: building graph: GraphTypePlan
2017/05/18 14:45:59 [WARN] terraform: shadow graph disabled
2017/05/18 14:45:59 [WARN] Test: Step plan: DIFF:

CREATE: librato_metric.foobar
  attributes.#:                 "0" => "1"
  attributes.0.display_stacked: "" => "true"
  composite:                    "" => "s(\"librato.cpu.percent.user\", {\"environment\" : \"prod\", \"service\": \"api\"})"
  description:                  "" => "A test composite metric"
  name:                         "" => "tftest-metric-nqvxkppkjs"
  type:                         "" => "composite"

STATE:

<no state>
2017/05/18 14:45:59 [INFO] terraform: building graph: GraphTypeApply
2017/05/18 14:45:59 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypePlan
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypeRefresh
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [INFO] Reading Librato Metric: tftest-metric-nqvxkppkjs
2017/05/18 14:46:00 [INFO] Read Librato Metric: {"name":"tftest-metric-nqvxkppkjs","description":"A test composite metric","type":"composite","composite":"s(\"librato.cpu.percent.user\", {\"environment\" : \"prod\", \"service\": \"api\"})","attributes":{"color":null,"display_max":null,"display_min":null,"display_units_long":"","display_units_short":"","display_stacked":true,"display_transform":"","created_by_ua":"go-librato/0.1"}}
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypePlan
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [WARN] Test: Executing destroy step
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypeValidate
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypeRefresh
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [INFO] Reading Librato Metric: tftest-metric-nqvxkppkjs
2017/05/18 14:46:00 [INFO] Read Librato Metric: {"name":"tftest-metric-nqvxkppkjs","description":"A test composite metric","type":"composite","composite":"s(\"librato.cpu.percent.user\", {\"environment\" : \"prod\", \"service\": \"api\"})","attributes":{"color":null,"display_max":null,"display_min":null,"display_units_long":"","display_units_short":"","display_stacked":true,"display_transform":"","created_by_ua":"go-librato/0.1"}}
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypePlanDestroy
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [WARN] Test: Step plan: DIFF:

DESTROY: librato_metric.foobar

STATE:

librato_metric.foobar:
  ID = tftest-metric-nqvxkppkjs
  attributes.# = 1
  attributes.0.aggregate = false
  attributes.0.color = 
  attributes.0.created_by_ua = go-librato/0.1
  attributes.0.display_max = 
  attributes.0.display_min = 
  attributes.0.display_stacked = true
  attributes.0.display_units_long = 
  attributes.0.display_units_short = 
  attributes.0.gap_detection = false
  composite = s("librato.cpu.percent.user", {"environment" : "prod", "service": "api"})
  description = A test composite metric
  name = tftest-metric-nqvxkppkjs
  type = composite
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypeApply
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [INFO] Deleting Metric: tftest-metric-nqvxkppkjs
2017/05/18 14:46:00 [INFO] Verifying Metric tftest-metric-nqvxkppkjs deleted
2017/05/18 14:46:00 [INFO] GETing Metric tftest-metric-nqvxkppkjs
2017/05/18 14:46:00 [INFO] GET returned a 404 for Metric tftest-metric-nqvxkppkjs
2017/05/18 14:46:00 [INFO] I think Metric is deleted
*** testAccCheckLibratoMetricDestroy ID: tftest-metric-nqvxkppkjs
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypePlanDestroy
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypeRefresh
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
2017/05/18 14:46:00 [INFO] terraform: building graph: GraphTypePlanDestroy
2017/05/18 14:46:00 [WARN] terraform: shadow graph disabled
--- FAIL: TestAccLibratoCompositeMetric (0.92s)
	testing.go:280: Step 0 error: After applying this step and refreshing, the plan was not empty:
		
		DIFF:
		
		UPDATE: librato_metric.foobar
		  attributes.0.created_by_ua: "go-librato/0.1" => ""
		
		STATE:
		
		librato_metric.foobar:
		  ID = tftest-metric-nqvxkppkjs
		  attributes.# = 1
		  attributes.0.aggregate = false
		  attributes.0.color = 
		  attributes.0.created_by_ua = go-librato/0.1
		  attributes.0.display_max = 
		  attributes.0.display_min = 
		  attributes.0.display_stacked = true
		  attributes.0.display_units_long = 
		  attributes.0.display_units_short = 
		  attributes.0.gap_detection = false
		  composite = s("librato.cpu.percent.user", {"environment" : "prod", "service": "api"})
		  description = A test composite metric
		  name = tftest-metric-nqvxkppkjs
		  type = composite
FAIL
exit status 1
FAIL	github.com/hashicorp/terraform/builtin/providers/librato	0.941s
make: *** [testacc] Error 1
```